projects/js-packages/critical-css-gen/package.json

```diff
@@ -57,6 +57,19 @@
 			"import": "./build-node/node.js",
 			"require": "./build-node/node.js",
 			"default": "./build-node/node.js"
+		},
+		"./node": {
+			"jetpack:src": "./src/node.ts",
+			"types": "./build-node/node.d.ts",
+			"import": "./build-node/node.js",
+			"require": "./build-node/node.js",
+			"default": "./build-node/node.js"
+		},
+		"./browser": {
+			"jetpack:src": "./src/browser.ts",
+			"import": "./build-browser/bundle.js",
+			"require": "./build-browser/bundle.js",
```

Comment:
This bundle isn't set up to be imported or required either. It's set up to be used something like
```html
<script src="build-browser/bundle.js"></script>
<script>
// a global "CriticalCSSGenerator" is defined by the above
CriticalCSSGenerator.whatever();
</script>
```

Reply:
We haven't switched to webpack yet 😅

From what I understand about our usage of the package, we're just using type declarations from it:
- https://github.com/Automattic/jetpack/blob/45cb9e1b6819501232e1de3dea0db62e8425bb12/projects/plugins/boost/app/assets/src/js/global.d.ts#L5-L8
- https://github.com/Automattic/jetpack/blob/45cb9e1b6819501232e1de3dea0db62e8425bb12/projects/plugins/boost/app/assets/src/js/features/critical-css/lib/generate-critical-css.ts#L9

The main work is done by a global variable coming from the browser build we're enqueueing in the Boost UI:

https://github.com/Automattic/jetpack/blob/45cb9e1b6819501232e1de3dea0db62e8425bb12/projects/plugins/boost/app/assets/src/js/features/critical-css/lib/generate-critical-css.ts#L211-L224

So I think I agree when you say that we don't need a browser endpoint, but I'm not sure how to get the type declarations from [`browser.ts`](https://github.com/Automattic/jetpack/blob/99baf6dfba29881377085a9ed375f611bfe60ac1/projects/js-packages/critical-css-gen/src/browser.ts).

Reply:
I thought #35079 switched Boost to Webpack?

> The main work is done by a global variable coming from the browser build we're enqueueing in the Boost UI

Yeah, that's the part I was suggesting changing. Let Webpack in Boost bundle critical-css-gen instead of relying on the global set by the standalone bundle.

You'd wind up changing the code in that linked file to `import` what it needs from the package instead of accessing it via the global, and the types should naturally come along with that.

Reply:
> I thought https://github.com/Automattic/jetpack/pull/35079 switched Boost to Webpack?

Oh man I got confused, apologies.

> Yeah, that's the part I was suggesting changing. Let Webpack in Boost bundle critical-css-gen instead of relying on the global set by the standalone bundle.

> You'd wind up changing the code in that linked file to import what it needs from the package instead of accessing it via the global, and the types should naturally come along with that.

Yeah, that would be ideal, except, we don't want to bundle the critical-css-gen code inside the boost bundle, since it's only used in the free version. Premium doesn't need it, as our cloud handles critical css generation.

Reply:
That makes sense, save a megabyte in size for paying customers if they don't need it.

Could you use a [dynamic import](https://webpack.js.org/guides/code-splitting/#dynamic-imports) instead of the global, e.g. `const { stuff } = await import( '@automattic/jetpack-critical-css-gen' );`? That has Webpack create a separate JS file, which isn't loaded in the browser until the code executes that `import()` statement.

Reply:
I've updated the code to use dynamic imports in 8d1294548a485814be8a3c9bfd4d847904a3a7f7.

I had to add the browser entry point, as [`BrowserInterfaceIframe`](https://github.com/Automattic/jetpack/blob/99baf6dfba29881377085a9ed375f611bfe60ac1/projects/js-packages/critical-css-gen/src/browser.ts#L0-L1) is not included in the default entry point (in [`node.ts`](https://github.com/Automattic/jetpack/blob/99baf6dfba29881377085a9ed375f611bfe60ac1/projects/js-packages/critical-css-gen/src/node.ts#L0-L1)).

With that done, there's now a problem with the build - it errors out. Errors include:

```
ERROR in ../../../node_modules/.pnpm/clean-css@5.3.3/node_modules/clean-css/lib/options/inline-request.js 1:10-24
Module not found: Error: Can't resolve 'url' in '/Users/ppetrov/projects/jetpack/node_modules/.pnpm/clean-css@5.3.3/node_modules/clean-css/lib/options'

BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
	- add a fallback 'resolve.fallback: { "url": require.resolve("url/") }'
	- install 'url'
If you don't want to include a polyfill, you can use an empty module like this:
	resolve.fallback: { "url": false }
```

```
ERROR in ../../../node_modules/.pnpm/clean-css@5.3.3/node_modules/clean-css/lib/options/rebase-to.js 1:11-26
Module not found: Error: Can't resolve 'path' in '/Users/ppetrov/projects/jetpack/node_modules/.pnpm/clean-css@5.3.3/node_modules/clean-css/lib/options'

BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
	- add a fallback 'resolve.fallback: { "path": require.resolve("path-browserify") }'
	- install 'path-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
	resolve.fallback: { "path": false }
```

```
ERROR in ../../../node_modules/.pnpm/clean-css@5.3.3/node_modules/clean-css/lib/reader/apply-source-maps.js 1:9-22
Module not found: Error: Can't resolve 'fs' in '/Users/ppetrov/projects/jetpack/node_modules/.pnpm/clean-css@5.3.3/node_modules/clean-css/lib/reader'
```

And a heap of other related to the `clean-css` package.

I tried adding pollyfills like the docs suggest:

> import() calls use [promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) internally. If you use import() with older browsers (e.g., IE 11), remember to shim Promise using a polyfill such as [es6-promise](https://github.com/stefanpenner/es6-promise) or [promise-polyfill](https://github.com/taylorhakes/promise-polyfill).

But that breaks things even worse.

I feel like I'm going down a rabbit hole and no matter which direction I take, something breaks. I'm wondering if using a dynamic import isn't an option (because of Boost's setup or something) and if we should just use a browser entry point and use the global.

Reply:
> Thanks for being patient with me. I really appreciate it.

You're welcome. Helping with the monorepo build tooling is what Garage is here for. 🙂

> I had to add the browser entry point, as [BrowserInterfaceIframe](https://github.com/Automattic/jetpack/blob/99baf6dfba29881377085a9ed375f611bfe60ac1/projects/js-packages/critical-css-gen/src/browser.ts#L0-L1) is not included in the default entry point (in [node.ts](https://github.com/Automattic/jetpack/blob/99baf6dfba29881377085a9ed375f611bfe60ac1/projects/js-packages/critical-css-gen/src/node.ts#L0-L1)).

The `./browser` entry point you have still only works within the monorepo.

What you should probably do is add `src/browser.ts` to tsconfig `include` ([here](https://github.com/Automattic/jetpack/blob/37326eb9216abc8496f0f8ba910ccfcdcdc6dbca/projects/js-packages/critical-css-gen/tsconfig.json#L3)), which will produce files at `./build-node/browser.js` and `./build-node/browser.d.ts` that the entry point should point to.

If you want you might also rename that `build-node/` dir to `build-module/` to be a bit more accurate. 🙂

Another alternative might be to switch things around, have `./build-node/browser.js` be the default entry point and have an entry point just for that `browser-interface-playwright.js` file.

> And a heap of other related to the `clean-css` package.

Hmm. `clean-css` doesn't seem to be very "clean" for bundling in the browser. :disappointed: 

> I tried fixing that by updating Boost's resolve section in webpack.config.js to:

That's probably the right track. I wonder how many of those are really needed though?

On my not-very-complicated test site, it worked with all of them set to `false` except for `path`. But maybe a site with more complex CSS would need some of the others? 🤷

I note that polyfills for `http` and `https` might be avoided by passing [a custom `fetch` handler](https://www.npmjs.com/package/clean-css#fetch-option) to `clean-css`.

>  but the build doesn't seem right

How so?

> and it also errors out when I try to generate critical css generation in the UI `process is not defined` error.

Looks like `clean-css` is assuming Node's `process` global variable is always present. To work around that, we'll need to `const webpack = require( 'webpack' );` at the top of the webpack config, add a dep on [process](https://www.npmjs.com/package/process) for a polyfill, and add this to the `plugins` to inject the global where it's needed:
```js
new webpack.ProvidePlugin( {
    process: require.resolve( 'process/browser' ),
} ),
```

Reply:
@anomiex before we move forward with your suggestion, I wanted to run something by you.

Today during a call with @haqadn, he suggested that we reduce the package to a single entry point and move project specific code to the respective projects.

So this would go to the Boost Cloud:

https://github.com/Automattic/jetpack/blob/199f5062ed30855d6344746b669a831fc5289a19/projects/js-packages/critical-css-gen/src/browser-interface-playwright.ts

And this would go to the Boost plugin:

https://github.com/Automattic/jetpack/blob/99baf6dfba29881377085a9ed375f611bfe60ac1/projects/js-packages/critical-css-gen/src/browser-interface-iframe.ts

That would mean, that we would no longer have `build-browser` or `build-node`, but just one build.

So the main file would probably look like this:

```js
export { BrowserInterface } from './browser-interface.js';
export { generateCriticalCSS } from './generate-critical-css.js';

export * from './errors.js';

export const version = '0.0.11';
```

What do you think about this idea? That sounds like something you might have suggested when we were migrating the package to the monorepo, I apologize if I didn't understand the idea initially.

Reply:
I have no objection to that refactor, if that makes sense to you who actually use and maintain the package. 🙂 That would also allow for dropping the optional peer dependency on playwright-core.

I also think dropping `build-browser/` would be good. It strikes me as a bit weird to supply a build that only works as `<script src="build-browser/bundle.js"></script>` inside the package. If an end user wants that, they should be able to make it happen better themself.

Even if you wanted to keep `BrowserInterfaceIframe` and/or `BrowserInterfacePlaywright` in the package, you could still drop `build-browser/`. You could also change the main file as suggested, and let people use `./browser-interface-iframe` and/or `./browser-interface-playwright` entry points if they want either of those classes.

Reply:
> I have no objection to that refactor, if that makes sense to you who actually use and maintain the package. 🙂 

To be honest, I'm not sure what makes the most sense, because the way the package is used in Boost is confusing to me. That's why I have a hard time figuring out the best approach.

I think I am getting a bit lost with the possible paths and their outcomes.

For the time being, I've decided to continue on what we were working towards previously.

The changes you suggested in https://github.com/Automattic/jetpack/pull/39509#discussion_r1778964527 have been implemented. I've also included some more changes.

The package build has been updated and the entry points have been changed.

- the rollup build (for the browser bundle) has been removed;
- browser (iframe interface) is now the default entry point (and has types, thanks!);
- there's now a `/playwright` entry point (to be used later in the boost cloud).

Everything seems to be working properly! However, the Boost build generated 4 new files:
- js-packages_critical-css-gen_src_browser_ts.js
- js-packages_critical-css-gen_src_browser_ts.js.map
- vendors-node_modules_pnpm_clean-css_5_3_3_node_modules_clean-css_index_js-node_modules_pnpm_c-7ef117.js
- vendors-node_modules_pnpm_clean-css_5_3_3_node_modules_clean-css_index_js-node_modules_pnpm_c-7ef117.js.map

I imagine those are because of the dynamic imports.

I am a bit worried about [this](https://github.com/Automattic/jetpack/pull/39509/commits/4603a38965a55ad296e593ae495b053817244a13#diff-ed7c252dcfd2e7ed286d669d0a94860deb9b0b8ca9aad9e5ab22a4b2a05d4a37R76-R87) part though:

```js
path: require.resolve( 'path-browserify' ),
process: require.resolve( 'process/browser' ),
url: false,
https: false,
http: false,
os: false,
buffer: false,
events: false,
fs: false,
```

Having all these here just to make the build happy seems like a band-aid and what you said earlier worries me a bit:

> But maybe a site with more complex CSS would need some of the others?

I wonder how this can be tested properly... I guess the rollup build was taking care of this previously 🤷🏻

Tests for the iframe interface also seem unhappy. Previously we were loading the bundle on the pages, but since there's no longer a bundle in the package, I'm not sure how to properly test [the iframe interface](https://github.com/Automattic/jetpack/blob/323a7b3ebbc82141becfb6746957d23f52dd9e94/projects/js-packages/critical-css-gen/tests/unit/browser-interface-iframe.test.js). Any ideas?

Reply:
> The package build has been updated and the entry points have been changed.
>
> * the rollup build (for the browser bundle) has been removed;
> * browser (iframe interface) is now the default entry point (and has types, thanks!);
> * there's now a /playwright entry point (to be used later in the boost cloud).

Sounds good to me!

> Everything seems to be working properly! However, the Boost build generated 4 new files:
>
> * js-packages_critical-css-gen_src_browser_ts.js
> * js-packages_critical-css-gen_src_browser_ts.js.map
> * vendors-node_modules_pnpm_clean-css_5_3_3_node_modules_clean-css_index_js-node_modules_pnpm_c-7ef117.js
> * vendors-node_modules_pnpm_clean-css_5_3_3_node_modules_clean-css_index_js-node_modules_pnpm_c-7ef117.js.map
>
> I imagine those are because of the dynamic imports.

That's right. In a production build they'll be named something like `543.js` and `931.js` instead.

If you want to rename the `js-packages_critical-css-gen_src_browser_ts.js` / `543.js` to something more readable, you could change the two `import( '@automattic/jetpack-critical-css-gen' )` to `import( /* webpackChunkName: "critical-css-gen" */ '@automattic/jetpack-critical-css-gen' )`. That'll also mean it won't get some other random number in production builds at some point in the future.

If you want to omit the separate file for `vendors-node_modules_pnpm_clean-css_5_3_3_node_modules_clean-css_index_js-node_modules_pnpm_c-7ef117.js` / `931.js` (merging it into the other one), in [the appropriate section of the Webpack config](https://github.com/Automattic/jetpack/blob/323a7b3ebbc82141becfb6746957d23f52dd9e94/projects/plugins/boost/webpack.config.js#L32-L34) you could adjust [`.optimization.splitChunks`](https://webpack.js.org/plugins/split-chunks-plugin/#optimizationsplitchunks), either setting it false or setting `.optimization.splitChunks.minChunks` to 2. Webpack's default settings seem to be aiming at a case where dynamic-loading two smaller files in parallel could be faster than one combined file.

> Having all these here just to make the build happy seems like a band-aid and what you said earlier worries me a bit
> [...]
> I wonder how this can be tested properly... I guess the rollup build was taking care of this previously 🤷🏻

Yeah, the `rollup-plugin-polyfill-node` plugin you had in there was presumably adding all these polyfills.

Without really testing, here's my take on the polyfills:

* fs: If anything tries to use the `fs` module, it's probably already broken. So setting it false seems like it should be ok. If something does break, changing CleanCSS settings to disable whatever it's trying to do with `fs` seems like it'd be the better solution.
* url: As far as I can tell this might be used for handling `@import` or applying source maps or url rebasing. But I think the default settings for all of those things is off?
* http, https: Same as url. If these did get used, I think the ideal thing would be to have critical-css-gen pass [a custom fetch handler](https://www.npmjs.com/package/clean-css#fetch-option) to `new CleanCSS()` [here](https://github.com/Automattic/jetpack/blob/update/boost/use-monorepo-css-gen-package/projects/js-packages/critical-css-gen/src/minify-css.ts#L14) instead of adding polyfills.
* os, buffer, events: Not needed with the current configs. Looks like setting `fs` false somehow removes the need for `os`, and `buffer` and `events` are only needed by the polyfills for `http` and/or `https`.

Reply:
> If you want to rename the js-packages_critical-css-gen_src_browser_ts.js / 543.js to something more readable, you could change the two import( '@automattic/jetpack-critical-css-gen' ) to import( /* webpackChunkName: "critical-css-gen" */ '@automattic/jetpack-critical-css-gen' ). That'll also mean it won't get some other random number in production builds at some point in the future.

This is awesome! I was wondering if there's a way to rename these, but didn't know what to check for in the docs. Updated in 533bf01c392125ef14f92d595c7a1086443d67ae.

> If you want to omit the separate file for vendors-node_modules_pnpm_clean-css_5_3_3_node_modules_clean-css_index_js-node_modules_pnpm_c-7ef117.js / 931.js (merging it into the other one), in [the appropriate section of the Webpack config](https://github.com/Automattic/jetpack/blob/323a7b3ebbc82141becfb6746957d23f52dd9e94/projects/plugins/boost/webpack.config.js#L32-L34) you could adjust [.optimization.splitChunks](https://webpack.js.org/plugins/split-chunks-plugin/#optimizationsplitchunks), either setting it false or setting .optimization.splitChunks.minChunks to 2. Webpack's default settings seem to be aiming at a case where dynamic-loading two smaller files in parallel could be faster than one combined file.

I've included this as well in e4d5a0f20ceb5cf3b0386139c15a26303e9a1539. I kind of like having just one file for this (like it was before).

> Yeah, the rollup-plugin-polyfill-node plugin you had in there was presumably adding all these polyfills.

Ah, knew it. Thanks for confirming.

> Without really testing, here's my take on the polyfills:

I did some tests for `@import` specifically. Interesting thing - they are omitted. I checked the source code for clean-css and it seemed to me that the code using url or http was disabled by default. I think we're good on that front.

> The `page.evaluate()` call, e.g. ... runs the passed function inside the "browser". So the import at the top of that file doesn't matter, what matters is what's loaded into the "page"

Figured as much. Was trying to load the `browser.js` (we get from the default package build) onto the browser, but I guess that wasn't the way to go. I wasn't sure if a separate build was necessary, but I guess it was 😅 

With the webpack config you provided I managed to get a build and update the tests to pass!

I am so happy with how this turned out.